File: MaraTrackDemo/MaraTrackDemo/RunViewController.m
//
//  RunViewController.m
//  MaraTrackDemo
//
//  Created by pk on 2017/8/8.
//  Copyright © 2017年 com.maramara. All rights reserved.
//

#import "RunViewController.h"
#import "Utils.h"

@interface RunViewController ()<MaraTrackManagerDelegate>

@property (weak, nonatomic) IBOutlet UILabel *distLabel;
@property (weak, nonatomic) IBOutlet UILabel *avrPaceLabel;
@property (weak, nonatomic) IBOutlet UILabel *curPaceLabel;
@property (weak, nonatomic) IBOutlet UILabel *durationLabel;
@property (weak, nonatomic) IBOutlet UIButton *startPauseBtn;
@property (weak, nonatomic) IBOutlet UILabel *statusLabel;

@property (nonatomic, strong) MaraTrackManager *trackManager;
@property (nonatomic, strong) TrackTask *tipsTask;
@end

@implementation RunViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    [self.navigationController setNavigationBarHidden:YES];
    MaraTrackerConfig *config = [[MaraTrackerConfig alloc] init];
    self.trackManager = [[MaraTrackManager alloc] initWith:config locManager:self.locationManager delegate:self];
    [self.trackManager startRun];
}

- (void)setTipsTask:(TrackTask *)tipsTask {
    _tipsTask = tipsTask;
    self.distLabel.text = [NSString stringWithFormat:@"%0.2f", floor(tipsTask.distance/10)/100];
    self.durationLabel.text = [Utils formatTime:tipsTask.time];
    self.avrPaceLabel.text = [Utils formatSpeed:tipsTask.avrPace];
    self.curPaceLabel.text = [Utils formatSpeed:tipsTask.curPace];
}
- (IBAction)startOrPauseDidClicked:(UIButton *)sender {
    // true 继续, false 暂停
    if (sender.isSelected) {
        [self.trackManager resumeRun];
    } else {
        [self.trackManager pauseRun];
    }
    [sender setSelected:!sender.isSelected];
}
- (IBAction)stopDidClicked:(UIButton *)sender {
    [self.trackManager stopRun];
    NSString *trackInfo = [self.trackManager getRunDetailJson];
    NSLog(@"trackInfo = %@", trackInfo);
    [self.navigationController popViewControllerAnimated:YES];
}

#pragma mark -- MaraTrackManagerDelegate

- (void)maraTrackManager:(MaraTrackManager *)manager didUpdateData:(TrackTask *)task {
    self.tipsTask = task;
}

- (void)maraTrackManager:(MaraTrackManager *)manager didUpdateStatus:(TrackerStatus)status {
    if (status == TrackerStatusAutoPaused) {
        self.statusLabel.text = @"自动暂停";
    } else if (status == TrackerStatusPaused) {
        self.statusLabel.text = @"手动暂停";
    } else if (status == TrackerStatusRun) {
        self.statusLabel.text = @"跑步中";
    }
}

- (void)maraTrackManager:(MaraTrackManager *)manager didUpdateLocation:(MaraUserLocation *)location {

}

- (void)maraTrackManager:(MaraTrackManager *)manager restoreInterruptData:(NSString *)json {
    [self.startPauseBtn setSelected:YES];
    self.statusLabel.text = @"恢复跑步,点击继续";
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];

}

@end

File: MaraTrackDemo/MaraTrackDemo-swift/RunViewController.swift
//
//  RunViewController.swift
//  MaraTrackDemo
//
//  Created by pk on 2017/8/15.
//  Copyright © 2017年 com.maramara. All rights reserved.
//

import UIKit
import MaraTrackSDK

class RunViewController: UIViewController {

    @IBOutlet weak var distLabel: UILabel!
    @IBOutlet weak var avrPaceLabel: UILabel!
    @IBOutlet weak var curPaceLabel: UILabel!
    @IBOutlet weak var timeLabel: UILabel!
    @IBOutlet weak var startPauseBtn: UIButton!
    @IBOutlet weak var statusLabel: UILabel!

    fileprivate var trackManager: MaraTrackManager?
    var locationManager: MaraLocationManager?
    
    fileprivate var tipsTask: TrackTask? {
        didSet {
            if let task = tipsTask {
                self.distLabel.text = String(format: "%0.2f", floor(task.distance/10)/100)
                self.timeLabel.text = Double(task.time).ext_hhmmssClockStringLimit()
                self.avrPaceLabel.text = task.avrPace.ext_mmssPaceString()
                self.curPaceLabel.text = task.curPace.ext_mmssPaceString()
            }
        }
    }

    override func viewDidLoad() {
        super.viewDidLoad()

        let config = MaraTrackerConfig()
        config.enableHA = true
        config.enableAutoPause = true
        self.trackManager = MaraTrackManager(config, locManager: self.locationManager, delegate: self)
        self.trackManager?.startRun()
    }

    @IBAction func startOrPauseDidClicked(_ sender: UIButton) {
        /// true 继续, false 暂停
        if sender.isSelected == true {
            self.trackManager?.resumeRun()
//            self.statusLabel.text = "跑步中";
        } else {
            self.trackManager?.pauseRun()
//            self.statusLabel.text = "手动暂停";
        }
        sender.isSelected = !sender.isSelected
    }

    @IBAction func stopDidClicked(_ sender: UIButton) {
        self.trackManager?.stopRun()
        if let trackInfo = self.trackManager?.getRunDetailJson() {
            debugPrint("trackinfo = \(trackInfo)")
        }
        self.navigationController?.popViewController(animated: true)
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()

    }

}

extension RunViewController: MaraTrackManagerDelegate {

    func maraTrackManager(_ manager: MaraTrackManager, didUpdateData task: TrackTask) {
        self.tipsTask = task
    }

    func maraTrackManager(_ manager: MaraTrackManager, didUpdate status: TrackerStatus) {
        if (status == .autoPaused) {
            self.statusLabel.text = "自动暂停";
            debugPrint("自动暂停")
        } else if (status == .paused) {
            self.statusLabel.text = "手动暂停";
        } else if (status == .run) {
            self.statusLabel.text = "跑步中";
        }
    }

    func maraTrackManager(_ manager: MaraTrackManager, didUpdate location: MaraUserLocation) {

    }

    func maraTrackManager(_ manager: MaraTrackManager, restoreInterruptData json: String) {
        self.startPauseBtn.isSelected = true
        self.statusLabel.text = "恢复跑步,点击继续";
    }
}
